src/groups/heisenberg.jl

```diff
@@ -0,0 +1,255 @@
+            "The matrix $(p) does not lie on $(G), since p[2:(n+1), 2:(n+1)] is not an identity matrix.",
+        )
+    end
+
+    return nothing
+end
+
+function check_vector(G::HeisenbergGroup{n}, p, X; kwargs...) where {n}
+    if !iszero(X[1, 1])
+        return DomainError(
+            X[1, 1],
+            "The matrix $(X) does not lie in the tangent space of $(G), since X[1, 1] is not equal to 0.",
+        )
+    end
+    if !iszero(X[n + 2, n + 2])
+        return DomainError(
+            X[n + 2, n + 2],
+            "The matrix $(X) does not lie in the tangent space of $(G), since X[n+2, n+2] is not equal to 0.",
+        )
+    end
+    if !iszero(X[2:(n + 2), 1:(n + 1)])
+        return DomainError(
+            norm(X[2:(n + 2), 1:(n + 1)]),
+            "The matrix $(X) does not lie in the tangent space of $(G), since X[2:(n + 2), 1:(n + 1)] is not a zero matrix.",
+        )
+    end
+    return nothing
+end
+
+embed(::HeisenbergGroup, p) = p
+embed(::HeisenbergGroup, p, X) = X
+
+@doc raw"""
+    get_coordinates(M::HeisenbergGroup, p, X, ::DefaultOrthonormalBasis{ℝ,TangentSpaceType})
+
+Get coordinates of tangent vector `X` at point `p` from the [`HeisenbergGroup`](@ref) `M`.
+Given a matrix
+```math
+\begin{bmatrix} 1 & \mathbf{a} & c \\
+\mathbf{0} & I_n & \mathbf{b} \\
+0 & \mathbf{0} & 1 \end{bmatrix}
+```
+the coordinates are concatenated vectors ``a``, ``b``, and number ``c``.
+"""
+get_coordinates(::HeisenbergGroup, p, X, ::DefaultOrthonormalBasis{ℝ,TangentSpaceType})
+
+function get_coordinates_orthonormal(M::HeisenbergGroup{n}, p, X, ::RealNumbers) where {n}
+    return vcat(_heisenberg_a_view(M, X), _heisenberg_b_view(M, X), X[1, n + 2])
+end
+
+function get_coordinates_orthonormal!(
+    M::HeisenbergGroup{n},
+    Xⁱ,
+    p,
+    X,
+    ::RealNumbers,
+) where {n}
+    Xⁱ[1:n] .= _heisenberg_a_view(M, X)
+    Xⁱ[(n + 1):(2 * n)] .= _heisenberg_b_view(M, X)
+    Xⁱ[2 * n + 1] = X[1, n + 2]
+    return Xⁱ
+end
+
+get_embedding(::HeisenbergGroup{n}) where {n} = Euclidean(n + 2, n + 2)
+
+function get_vector_orthonormal(::HeisenbergGroup{n}, p, Xⁱ, ::RealNumbers) where {n}
+    return [
+        0 Xⁱ[1:n] Xⁱ[2 * n + 1]
+        zeros(n, n + 1) Xⁱ[(n + 1):(2 * n)]'
+        zeros(1, n + 2)
+    ]
+end
+
+function get_vector_orthonormal!(::HeisenbergGroup{n}, X, p, Xⁱ, ::RealNumbers) where {n}
+    fill!(X, 0)
+    X[1, 2:(n + 1)] .= Xⁱ[1:n]
+    X[2:(n + 1), n + 2] .= Xⁱ[(n + 1):(2 * n)]
+    X[1, n + 2] = Xⁱ[2 * n + 1]
+    return X
+end
+
+function exp_lie!(M::HeisenbergGroup{n}, q, X) where {n}
+    copyto!(q, I)
+    a_view = _heisenberg_a_view(M, X)
+    b_view = _heisenberg_b_view(M, X)
+    q[1, 2:(n + 1)] .= a_view
+    q[2:(n + 1), n + 2] .= b_view
+    q[1, n + 2] = X[1, n + 2] + dot(a_view, b_view) / 2
+    return q
+end
+
+@doc raw"""
+    exp(M::HeisenbergGroup, p, X)
+
+Exponential map on the [`HeisenbergGroup`](@ref) `M` with the left-invariant metric.
+"""
```

Comment:
can we maybe add a formula here? I think that would be nice.

Reply:
I'm not exactly sure what to write here. I took the standard definition for exponential at identity and extended it through left-invariance (hence `p \ q`).

Reply:
I see that in the code, but also the lie exp is not documented, so I would love to have a formula here to at exp_lie

Reply:
I've added a formula.